smart-contracts/contracts/mixins/MixinPurchase.sol

```diff
@@ -57,94 +57,106 @@ contract MixinPurchase is
   */
   function purchase(
     uint256 _value,
-    address _recipient,
-    address _referrer,
-    address _keyManager,
+    address[] memory _recipients,
+    address[] memory _referrers,
+    address[] memory _keyManagers,
     bytes calldata _data
   ) external payable
     onlyIfAlive
     notSoldOut
   {
-    require(_recipient != address(0), 'INVALID_ADDRESS');
-
-    // Assign the key
-    Key storage toKey = keyByOwner[_recipient];
-    uint idTo = toKey.tokenId;
-    uint newTimeStamp;
-
-    if (idTo == 0) {
-      // Assign a new tokenId (if a new owner or previously transferred)
-      _assignNewTokenId(toKey);
-      // refresh the cached value
-      idTo = toKey.tokenId;
-      _recordOwner(_recipient, idTo);
-      // check for a non-expiring key
-      if (expirationDuration == type(uint).max) {
-        newTimeStamp = type(uint).max;
-      } else {
-        newTimeStamp = block.timestamp + expirationDuration;
-      }
-      toKey.expirationTimestamp = newTimeStamp;
-
-      // set key manager
-      _setKeyManagerOf(idTo, _keyManager);
-
-      // trigger event
-      emit Transfer(
-        address(0), // This is a creation.
-        _recipient,
-        idTo
-      );
-    } else if (toKey.expirationTimestamp > block.timestamp) {
-      // prevent re-purchase of a valid non-expiring key
-      require(toKey.expirationTimestamp != type(uint).max, 'A valid non-expiring key can not be purchased twice');
-
-      // This is an existing owner trying to extend their key
-      newTimeStamp = toKey.expirationTimestamp + expirationDuration;
-      toKey.expirationTimestamp = newTimeStamp;
-
-      emit RenewKeyPurchase(_recipient, newTimeStamp);
-    } else {
-      // This is an existing owner trying to renew their expired or cancelled key
-      if(expirationDuration == type(uint).max) {
-        newTimeStamp = type(uint).max;
+    require(_recipients.length == _referrers.length, 'INVALID_REFERRERS_LENGTH');
+    require(_recipients.length == _keyManagers.length, 'INVALID_KEY_MANAGERS_LENGTH');
+
+    uint totalPriceToPay;
+
+    for (uint256 i = 0; i < _recipients.length; i++) {
+      // check recipient address
+      address _recipient = _recipients[i];
+      require(_recipient != address(0), 'INVALID_ADDRESS');
```

Comment:
I actually think we could lift this requirement since the key manager could transfer that key anyway.

Reply:
but what if the keyManager is invalid too?  

In practice, it is likely the same array will be passed for both recipients and keyManagers. If the mistake is in both then there are no ways to revert it and the key will be lost.